uda_udf_test.py

```diff
@@ -0,0 +1,111 @@
+        self.log.info("Prewriting database...")
+        stress_cmd = self.params.get('prepare_write_cmd')
+        pre_thread = self.run_stress_thread(stress_cmd=stress_cmd, stats_aggregate_cmds=False, round_robin=False)
+        self.verify_stress_thread(cs_thread_pool=pre_thread)
+        self.log.info("Database pre write completed")
+
+    def _verify_udf_functions(self):
+        row_query = UDFVerification(name="row_query",
+                                    query=f"SELECT * FROM {self.KEYSPACE_NAME}.{self.CF_NAME} LIMIT 1",
+                                    verifier_func=lambda c2, c3, c7: all([c2, c3, c7]))
+        verifications = [
+            UDFVerification(name="lua_var_length_counter",
+                            query=f"SELECT {self.KEYSPACE_NAME}.lua_var_length_counter(c7) AS result "
+                                  f"FROM {self.KEYSPACE_NAME}.{self.CF_NAME} LIMIT 1",
+                            verifier_func=lambda c2, c3, c7, query_response: len(c7) == query_response.result),
+            UDFVerification(name="xwasm_plus",
+                            query=f"SELECT {self.KEYSPACE_NAME}.xwasm_plus(c2, c3) AS result "
+                                  f"FROM {self.KEYSPACE_NAME}.{self.CF_NAME} LIMIT 1",
+                            verifier_func=lambda c2, c3, c7, query_response: c2 + c3 == query_response.result),
+            UDFVerification(name="xwasm_div",
+                            query=f"SELECT {self.KEYSPACE_NAME}.xwasm_div(c2, c3) AS result "
+                                  f"FROM {self.KEYSPACE_NAME}.{self.CF_NAME} LIMIT 1",
+                            verifier_func=lambda c2, c3, c7, query_response: c2 // c3 == query_response.result),
+            # UDFVerification(name="xwasm_fib",
+            #                 query=f"SELECT {self.KEYSPACE_NAME}.xwasm_fib(c2) AS result "
+            #                       f"FROM {self.KEYSPACE_NAME}.{self.CF_NAME} LIMIT 1",
+            #                 verifier_func=lambda c2, c3, c7, query_response: self._fib(c2) == query_response.result)
+        ]
+        self.log.info("Starting UDF verifications...")
+
+        with self.db_cluster.cql_connection_patient(self.db_cluster.get_node(), verbose=False) as session:
+            row_result = session.execute(row_query.query).one()
+            c2_value = row_result.c2
+            c3_value = row_result.c3
+            c7_value = row_result.c7
+            assert row_query.verifier_func(c2_value, c3_value, c7_value), \
+                "Expected row values to not be None, at least one them was. " \
+                "c2_value: %s, c3_value: %s, c7 value: %s" % (c2_value, c3_value, c7_value)
+
+            for verification in verifications:
+                self.log.info("Running UDF verification: %s; query: %s", verification.name, verification.query)
+                query_result = session.execute(verification.query).one()
+                assert verification.verifier_func(c2_value, c3_value, c7_value, query_result)
```

Comment:
how can you be sure that each verification query returns the same row as in `row_result` when there's no `where` clause in queries?

Reply:
Why would it not? The data is pre-written and doesn't change during the test - I'd expect the results to be consistent in such a situation. (Plus, the verifications are passing in my staging tests.)

Reply:
I meant it is not sorted with `WHERE` or `ORDER BY` - so there's no certainty it will result in the same row. I wonder how it works.

Reply:
The data is already sorted by partition (pk) and cluster keys (ck_int). From our docs[1] on the `CLUSTERING ORDER` param for table creation: `it also changes the default order of results when queried (if no ORDER BY is provided). Results are always returned in clustering order (within a partition).` In this case we stick to defaults, no CLUSTERING ORDER is defined, so the data will always be retrieved in the same order while there is nothing affecting the partitioning and clustering. 
Either way, I also double checked on a small toy example with creating different client session and redoing a `SELECT * LIMIT 1;` query and the results are stable.

[1] https://docs.scylladb.com/stable/cql/ddl.html#create-table-statement

Reply:
Thanks for the explanation @KnifeyMoloko